From 15dc2a31a4fcb840306e6de5ff8f1a5532324bb2 Mon Sep 17 00:00:00 2001
From: Chunquan Luo <chunquan@qti.qualcomm.com>
Date: Fri, 9 Oct 2026 01:56:17 -0700
Subject: [PATCH] qcacld-3.0: Fix MLO connect

MLO auth fail for mld addr convert to peer addr fail.

CRs-Fixed: 4695577
Change-Id: I74adb37985d9455d3ea5570899cad1b4432c7440
---
 .../dispatcher/inc/wlan_cm_roam_api.h         | 41 +++++-------
 .../inc/wlan_cm_roam_public_struct.h          |  2 +-
 .../dispatcher/inc/wlan_cm_roam_ucfg_api.h    |  2 +-
 .../dispatcher/src/wlan_cm_roam_api.c         | 67 ++++++++++---------
 core/mac/src/pe/lim/lim_process_auth_frame.c  |  9 ---
 .../src/pe/lim/lim_send_management_frames.c   |  2 +-
 6 files changed, 53 insertions(+), 70 deletions(-)

diff --git a/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_api.h b/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_api.h
index 1baa5d7b26..ad608b9909 100644
--- a/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_api.h
+++ b/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_api.h
@@ -2314,19 +2314,7 @@ wlan_cm_is_sae_auth_addr_conversion_required(struct wlan_objmgr_vdev *vdev)
 }
 #endif /* WLAN_FEATURE_11BE_MLO */
 
-#if defined(WLAN_FEATURE_11BE_MLO) && defined(WLAN_FEATURE_ROAM_OFFLOAD)
-/**
- * wlan_cm_store_mlo_roam_peer_address() - cache peer mld and link address
- * while roaming
- * @pdev: pdev object
- * @auth_event: auth offload event data
- *
- * Return: void
- */
-void
-wlan_cm_store_mlo_roam_peer_address(struct wlan_objmgr_pdev *pdev,
-				    struct auth_offload_event *auth_event);
-
+#ifdef WLAN_FEATURE_11BE_MLO
 /**
  * wlan_cm_roaming_get_peer_mld_addr() - retrieve the peer mld address for
  * roaming.
@@ -2345,6 +2333,21 @@ wlan_cm_roaming_get_peer_mld_addr(struct wlan_objmgr_vdev *vdev);
  */
 struct qdf_mac_addr *
 wlan_cm_roaming_get_peer_link_addr(struct wlan_objmgr_vdev *vdev);
+#endif
+
+#if defined(WLAN_FEATURE_11BE_MLO) && defined(WLAN_FEATURE_ROAM_OFFLOAD)
+/**
+ * wlan_cm_store_mlo_roam_peer_address() - cache peer mld and link address
+ * while roaming
+ * @pdev: pdev object
+ * @auth_event: auth offload event data
+ *
+ * Return: void
+ */
+void
+wlan_cm_store_mlo_roam_peer_address(struct wlan_objmgr_pdev *pdev,
+				    struct auth_offload_event *auth_event);
+
 
 /**
  * wlan_cm_roam_is_mlo_ap() - to check whether vdev will be roam to ml ap
@@ -2375,18 +2378,6 @@ wlan_cm_store_mlo_roam_peer_address(struct wlan_objmgr_pdev *pdev,
 {
 }
 
-static inline struct qdf_mac_addr *
-wlan_cm_roaming_get_mld_addr(struct wlan_objmgr_vdev *vdev)
-{
-	return NULL;
-}
-
-static inline struct qdf_mac_addr *
-wlan_cm_roaming_get_peer_link_addr(struct wlan_objmgr_vdev *vdev)
-{
-	return NULL;
-}
-
 static inline bool
 wlan_cm_roam_is_mlo_ap(struct wlan_objmgr_vdev *vdev)
 {
diff --git a/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_public_struct.h b/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_public_struct.h
index 4f31afe35e..5c11af6c69 100644
--- a/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_public_struct.h
+++ b/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_public_struct.h
@@ -666,7 +666,7 @@ struct rso_config {
 	int32_t lost_link_rssi;
 	struct roam_synch_frame_ind roam_sync_frame_ind;
 	uint32_t roam_band_bitmask;
-#if defined(WLAN_FEATURE_11BE_MLO) && defined(WLAN_FEATURE_ROAM_OFFLOAD)
+#ifdef WLAN_FEATURE_11BE_MLO
 	struct sae_roam_auth_map sae_roam_auth;
 #endif
 	enum wlan_cm_source roam_invoke_source;
diff --git a/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_ucfg_api.h b/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_ucfg_api.h
index c06888cf18..315f44e72e 100644
--- a/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_ucfg_api.h
+++ b/components/umac/mlme/connection_mgr/dispatcher/inc/wlan_cm_roam_ucfg_api.h
@@ -896,7 +896,7 @@ ucfg_cm_is_sae_auth_addr_conversion_required(struct wlan_objmgr_vdev *vdev)
 }
 #endif
 
-#if defined(WLAN_FEATURE_ROAM_OFFLOAD) && defined(WLAN_FEATURE_11BE_MLO)
+#ifdef WLAN_FEATURE_11BE_MLO
 /**
  * ucfg_cm_roaming_get_peer_mld_addr() - this api is wrapper for
  * "wlan_cm_roaming_get_peer_mld_addr" function.
diff --git a/components/umac/mlme/connection_mgr/dispatcher/src/wlan_cm_roam_api.c b/components/umac/mlme/connection_mgr/dispatcher/src/wlan_cm_roam_api.c
index 48700310ee..d91ebf4f8a 100644
--- a/components/umac/mlme/connection_mgr/dispatcher/src/wlan_cm_roam_api.c
+++ b/components/umac/mlme/connection_mgr/dispatcher/src/wlan_cm_roam_api.c
@@ -6013,9 +6013,39 @@ bool wlan_cm_is_sae_auth_addr_conversion_required(struct wlan_objmgr_vdev *vdev)
 
 	return true;
 }
-#endif /* WLAN_FEATURE_11BE_MLO */
 
-#if defined(WLAN_FEATURE_ROAM_OFFLOAD) && defined(WLAN_FEATURE_11BE_MLO)
+struct qdf_mac_addr *
+wlan_cm_roaming_get_peer_mld_addr(struct wlan_objmgr_vdev *vdev)
+{
+	struct rso_config *rso_config;
+
+	rso_config = wlan_cm_get_rso_config(vdev);
+	if (!rso_config)
+		return NULL;
+
+	if (qdf_is_macaddr_zero(&rso_config->sae_roam_auth.peer_mldaddr))
+		return NULL;
+
+	return &rso_config->sae_roam_auth.peer_mldaddr;
+}
+
+struct qdf_mac_addr *
+wlan_cm_roaming_get_peer_link_addr(struct wlan_objmgr_vdev *vdev)
+{
+	struct rso_config *rso_config;
+
+	rso_config = wlan_cm_get_rso_config(vdev);
+	if (!rso_config)
+		return NULL;
+
+	if (qdf_is_macaddr_zero(&rso_config->sae_roam_auth.peer_linkaddr))
+		return NULL;
+
+	return &rso_config->sae_roam_auth.peer_linkaddr;
+}
+#endif
+
+#if defined(WLAN_FEATURE_11BE_MLO) && defined(WLAN_FEATURE_ROAM_OFFLOAD)
 /**
  * wlan_cm_reset_mlo_roam_peer_address() - this API reset the sae_roam_auth
  * structure values to zero.
@@ -6090,36 +6120,6 @@ rel_ref:
 	wlan_objmgr_vdev_release_ref(vdev, WLAN_LEGACY_MAC_ID);
 }
 
-struct qdf_mac_addr *
-wlan_cm_roaming_get_peer_mld_addr(struct wlan_objmgr_vdev *vdev)
-{
-	struct rso_config *rso_config;
-
-	rso_config = wlan_cm_get_rso_config(vdev);
-	if (!rso_config)
-		return NULL;
-
-	if (qdf_is_macaddr_zero(&rso_config->sae_roam_auth.peer_mldaddr))
-		return NULL;
-
-	return &rso_config->sae_roam_auth.peer_mldaddr;
-}
-
-struct qdf_mac_addr *
-wlan_cm_roaming_get_peer_link_addr(struct wlan_objmgr_vdev *vdev)
-{
-	struct rso_config *rso_config;
-
-	rso_config = wlan_cm_get_rso_config(vdev);
-	if (!rso_config)
-		return NULL;
-
-	if (qdf_is_macaddr_zero(&rso_config->sae_roam_auth.peer_linkaddr))
-		return NULL;
-
-	return &rso_config->sae_roam_auth.peer_linkaddr;
-}
-
 bool wlan_cm_roam_is_mlo_ap(struct wlan_objmgr_vdev *vdev)
 {
 	struct rso_config *rso_config;
@@ -6144,8 +6144,9 @@ wlan_cm_add_all_link_probe_rsp_to_scan_db(struct wlan_objmgr_psoc *psoc,
 {
 	return mlo_add_all_link_probe_rsp_to_scan_db(psoc, candidate);
 }
+#endif /* WLAN_FEATURE_11BE_MLO */
 
-#elif defined(WLAN_FEATURE_ROAM_OFFLOAD) /* end WLAN_FEATURE_11BE_MLO */
+#ifdef WLAN_FEATURE_ROAM_OFFLOAD
 QDF_STATUS
 cm_roam_candidate_event_handler(struct wlan_objmgr_psoc *psoc,
 				struct roam_scan_candidate_frame *candidate)
diff --git a/core/mac/src/pe/lim/lim_process_auth_frame.c b/core/mac/src/pe/lim/lim_process_auth_frame.c
index d06a0de419..9f04110614 100644
--- a/core/mac/src/pe/lim/lim_process_auth_frame.c
+++ b/core/mac/src/pe/lim/lim_process_auth_frame.c
@@ -629,7 +629,6 @@ static void lim_get_sta_mld_address(struct wlan_objmgr_vdev *vdev,
 	util_get_bvmlie_mldmacaddr(ml_ie, ml_ie_total_len, peer_mld);
 }
 
-#ifdef WLAN_FEATURE_ROAM_OFFLOAD
 /**
  * lim_update_link_to_mld_address:This API is called to update SA and DA address
  * @mac_ctx: Pointer to mac context
@@ -700,14 +699,6 @@ static QDF_STATUS lim_update_link_to_mld_address(struct mac_context *mac_ctx,
 
 	return QDF_STATUS_SUCCESS;
 }
-#else
-static QDF_STATUS lim_update_link_to_mld_address(struct mac_context *mac_ctx,
-                                                 struct wlan_objmgr_vdev *vdev,
-                                                 tpSirMacMgmtHdr mac_hdr)
-{
-	return QDF_STATUS_SUCCESS;
-}
-#endif /* WLAN_FEATURE_ROAM_OFFLOAD */
 
 #else
 static void lim_get_sta_mld_address(struct wlan_objmgr_vdev *vdev,
diff --git a/core/mac/src/pe/lim/lim_send_management_frames.c b/core/mac/src/pe/lim/lim_send_management_frames.c
index 344cabc962..b1efb51e8b 100644
--- a/core/mac/src/pe/lim/lim_send_management_frames.c
+++ b/core/mac/src/pe/lim/lim_send_management_frames.c
@@ -8416,7 +8416,7 @@ lim_handle_sae_auth_retry(struct mac_context *mac_ctx, uint8_t vdev_id,
 	}
 }
 
-#if (defined WLAN_FEATURE_11BE_MLO) && (defined WLAN_FEATURE_ROAM_OFFLOAD)
+#ifdef WLAN_FEATURE_11BE_MLO
 static QDF_STATUS lim_update_mld_to_link_address(struct mac_context *mac_ctx,
 						 struct wlan_objmgr_vdev *vdev,
 						 tpSirMacMgmtHdr mac_hdr)
-- 
2.34.1

